1/src/Shape.ts

```diff
@@ -0,0 +1,43 @@
+import {Point} from "./point";
+
+export class Shape {
+  public color: string;
+  public filled: boolean;
+  public points: Point[];
+
+  constructor(points: Point[]);
+  constructor(points: Point[], color: string, filled: boolean);
+  constructor(points: Point[], color?: string, filled?: boolean) {
+    if (points.length < 3) {
+      throw new Error('Should be at least 3 points');
+    }
+    this.points = points;
+
+    if (!color && !filled) {
+      this.color = 'green';
+      this.filled = true;
+    } else {
+      this.color = color;
+      this.filled = filled;
+    }
+  }
+
+  public toString(): string {
+    const resultString: string = `A Shape with color of ${this.color} and ${this.filled ? 'filled' : 'not filled'}. Points:`
+    const points: string = this.points.reduce((pointsString: string, point: Point, index: number) => {
+      return `${pointsString} ${point.toString()}${(index === this.points.length - 1 ? '.' : ',')}`
+    }, '');
+    return resultString + points;
```

Comment:
You can create a separate function for creating points string and instead of "str1 + str2" add directly to your resultString. Ideally your `toString` method should have only a single return.

Reply:
You didn't fix this comment. Split points into different method and don't use string concatenation with +
Something like this
```typescript
return `Some text ${val1}, ${val2}, points - ${this.generatePointsString()}`;
```